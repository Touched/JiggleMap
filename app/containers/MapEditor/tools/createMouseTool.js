import { combineReducers } from 'redux';
import invariant from 'invariant';

export const MOUSE_PRESS = 'jigglemap/MapEditor/MouseTool/MOUSE_PRESS';
export const MOUSE = 'jigglemap/MapEditor/MouseTool/MOUSE';
export const MOUSE_RELEASE = 'jigglemap/MapEditor/MouseTool/MOUSE_RELEASE';

export const initialState = {
  startingPosition: null,
  currentPosition: null,
  clientOffset: null,
  currentBlock: 0,
  object: null,
};

export type Position = {
  x: number;
  y: number;
};

export type MouseToolState = {
  startingPosition: ?Position;
  currentPosition: ?Position;
  clientOffset: ?Position;
  currentBlock: number;
  object: ?Object;
};

export type MousePressAction = {
  type: typeof MOUSE_PRESS;
  clientOffset: Position;
  position: Position;
  object: Object;
};

export type MouseAction = {
  type: typeof MOUSE;
  position: Position;
};

export type MouseReleaseAction = {
  type: typeof MOUSE_RELEASE;
};

export type Action = MousePressAction | MouseAction | MouseReleaseAction;

export type CombinedState<State> = {
  mouse: MouseToolState;
  tool: State;
};

export type MouseTool<State> = {
  id: string;
  name: Message;
  description: Message;
  layers: Array<Layer>;
  icon: React.Element<*>;
  component: React.ComponentType<any>;
  reducer?: (state: CombinedState<State>, action: Action) => CombinedState<State>;
  getCursorForObject: (object: Object, mouseState: CombinedState, state: State) => string;
  onMousePress: (object: Object, position: Position, mouseState: MouseToolState, state: State, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) => void;
  onMouse: (position: Position, mouseState: MouseToolState, state: State, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) => void;
  onMouseRelease: (mouseState: MouseToolState, state: State, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) => void;
};

function toGridCoordinates(x, y) {
  return {
    x: Math.floor(x / 16),
    y: Math.floor(y / 16),
  };
}

function reducer(state: MouseToolState = initialState, action: Action): MouseToolState {
  switch (action.type) {
    case MOUSE_PRESS:
      return {
        ...state,
        clientOffset: action.clientOffset,
        startingPosition: action.position,
        currentPosition: action.position,
        object: action.object,
      };
    case MOUSE:
      return {
        ...state,
        currentPosition: action.position,
      };
    case MOUSE_RELEASE:
      return {
        ...state,
        startingPosition: null,
      };
    default:
      (action: empty); // eslint-disable-line no-unused-expressions
      return state;
  }
}

export default function createMouseTool(definition): MouseTool<CombinedState<State>> {
  type State = CombinedState<T>;

  const combinedReducer = combineReducers({
    mouse: reducer,
    tool: definition.reducer || ((state = {}) => state),
  });

  return {
    ...definition,
    type: 'mouse',
    getCursorForObject(object, state) {
      if (this.handlesType(object.type)) {
        return definition.getCursorForObject(object, state.mouse, state.tool);
      }

      return 'auto';
    },
    onMouseDown(object: Object, state: State, meta: Meta, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) {
      if (this.handlesType(object.type)) {
        const { nativeEvent: { offsetX, offsetY }, clientX, clientY } = mouseEvent;
        const position = toGridCoordinates(offsetX, offsetY);

        tabDispatch({
          type: MOUSE_PRESS,
          clientOffset: {
            x: (clientX / meta.zoomLevel) - offsetX,
            y: (clientY / meta.zoomLevel) - offsetY,
          },
          position,
          object,
        });

        const mouseState = {
          ...state.mouse,
          startingPosition: position,
        };

        this.onMousePress(position, mouseState, state.tool, tabDispatch, mouseEvent);
      }
    },
    onMouseMove(state: State, meta: Meta, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) {
      if (state.mouse.startingPosition) {
        const { clientOffset, currentPosition } = state.mouse;

        invariant(clientOffset, 'mousedown event did not set the clientOffset');
        invariant(currentPosition, 'mousedown event did not set the currentPosition');

        const x = (mouseEvent.clientX / meta.zoomLevel) - clientOffset.x;
        const y = (mouseEvent.clientY / meta.zoomLevel) - clientOffset.y;
        const position = toGridCoordinates(x, y);

        if (currentPosition.x !== position.x || currentPosition.y !== position.y) {
          tabDispatch({
            type: MOUSE,
            position,
          });

          this.onMouse(position, state.mouse, state.tool, tabDispatch, mouseEvent);
        }
      }
    },
    onMouseUp(state: State, meta: Meta, tabDispatch: Dispatch, mouseEvent: ReactMouseEvent) {
      if (state.mouse.startingPosition) {
        tabDispatch({
          type: MOUSE_RELEASE,
        });

        this.onMouseRelease(state.mouse, state.tool, tabDispatch, mouseEvent);
      }
    },
    reducer: combinedReducer,
  };
}
