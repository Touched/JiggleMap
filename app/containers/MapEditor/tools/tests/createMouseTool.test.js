import createMouseTool, { MOUSE_PRESS, MOUSE, MOUSE_RELEASE, initialState } from '../createMouseTool';

describe('createMouseTool', () => {
  let mouseTool;
  let getCursorForObject;
  const reducer = jest.fn((state = { foo: 'bar' }) => state);

  beforeEach(() => {
    getCursorForObject = jest.fn(() => 'pointer');
    mouseTool = createMouseTool({
      onMousePress: jest.fn(),
      onMouse: jest.fn(),
      onMouseRelease: jest.fn(),
      getCursorForObject,
      handlesType(type) {
        return type === 'main-map';
      },
      reducer,
    });
  });

  const object = {
    type: 'main-map',
  };

  const meta = {
    zoomLevel: 2,
  };

  const dispatch = jest.fn();

  it('returns a cursor only over the handle type', () => {
    expect(mouseTool.getCursorForObject({ type: 'main-map' }, {})).toEqual('pointer');
    expect(mouseTool.getCursorForObject({ type: 'other' }, {})).toEqual('auto');
  });

  it('calls getCursorForObject', () => {
    mouseTool.getCursorForObject({ type: 'main-map' }, { mouse: 'foo', tool: 'bar' });
    expect(getCursorForObject).toHaveBeenCalledWith({ type: 'main-map' }, 'foo', 'bar');
  });

  describe('reducer', () => {
    it('sets the intial state', () => {
      expect(mouseTool.reducer(undefined, {})).toMatchSnapshot();
      expect(createMouseTool({}).reducer(undefined, {})).toMatchSnapshot();
    });

    describe('MOUSE_PRESS', () => {
      it('stores the information', () => {
        const position = {
          x: 10,
          y: 5,
        };

        const clientOffset = { x: 156, y: 167 };

        const state = mouseTool.reducer({}, {
          type: MOUSE_PRESS,
          object,
          position,
          clientOffset,
        });

        expect(state).toMatchObject({
          mouse: {
            clientOffset,
            startingPosition: position,
            currentPosition: position,
            object,
          },
        });
      });
    });

    describe('MOUSE', () => {
      it('updates the currentPosition', () => {
        const position = { x: 10, y: 5 };
        const state = mouseTool.reducer({}, {
          type: MOUSE,
          position,
        });

        expect(state).toMatchObject({
          mouse: {
            currentPosition: position,
          },
        });
      });
    });

    describe('MOUSE_RELEASE', () => {
      it('sets the startingPosition to null', () => {
        const state = mouseTool.reducer({}, {
          type: MOUSE_RELEASE,
        });

        expect(state).toMatchObject({
          mouse: {
            startingPosition: null,
          },
        });
      });
    });

    describe('default', () => {
      it('passes the action through to the child reducer', () => {
        const action = {
          type: 'foo',
        };

        const state = mouseTool.reducer(undefined, action);

        expect(reducer).toHaveBeenCalledWith(undefined, action);
        expect(state).toEqual({
          mouse: initialState,
          tool: {
            foo: 'bar',
          },
        });
      });
    });
  });

  describe('mouse press', () => {
    const event = {
      nativeEvent: {
        offsetX: 160,
        offsetY: 80,
      },
      clientX: 456,
      clientY: 186,
    };

    const position = {
      x: 10,
      y: 5,
    };

    const state = {
      mouse: {
        foo: 'bar',
      },
      tool: {
        baz: 'quz',
      },
    };

    const updatedState = {
      ...state,
      mouse: {
        ...state.mouse,
        startingPosition: position,
      },
    };

    it('it does not call onMousePress when onMouseDown is called with a non main-map object', () => {
      mouseTool.onMouseDown({});
      expect(mouseTool.onMousePress).not.toHaveBeenCalled();
    });

    it('calls onMousePress when onMouseDown is called with a main-map object', () => {
      mouseTool.onMouseDown(object, state, meta, dispatch, event);
      expect(mouseTool.onMousePress).toHaveBeenCalledWith(position, updatedState.mouse, updatedState.tool, dispatch, event);
    });

    it('dispatches a MOUSE_PRESS action', () => {
      mouseTool.onMouseDown(object, updatedState, meta, dispatch, event);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: MOUSE_PRESS,
        clientOffset: {
          x: 68,
          y: 13,
        },
        position,
        object,
      });
    });
  });

  describe('mouse', () => {
    const event = {
      nativeEvent: {
        offsetX: 160,
        offsetY: 80,
      },
      clientX: 456,
      clientY: 186,
    };

    const position = {
      x: 10,
      y: 5,
    };

    const state = {
      mouse: {
        clientOffset: {
          x: 115,
          y: 76,
        },
        startingPosition: position,
        currentPosition: position,
      },
      tool: {
        foo: 'bar',
      },
    };

    it('does not call onMouse if there was no onMousePress', () => {
      mouseTool.onMouseMove({ mouse: {} }, meta, dispatch, event);
      expect(mouseTool.onMouse).not.toHaveBeenCalled();
    });

    it('calls onMouse if there was a onMousePress', () => {
      mouseTool.onMouseMove(state, meta, dispatch, event);
      expect(mouseTool.onMouse).toHaveBeenCalledWith({
        x: 7,
        y: 1,
      }, state.mouse, state.tool, dispatch, event);
    });

    it('does not call onMouse if the cursor has not moved to a different grid cell', () => {
      const newState = {
        ...state,
        mouse: {
          ...state.mouse,
          startingPosition: position,
          currentPosition: { x: 7, y: 1 },
        },
      };

      mouseTool.onMouseMove(newState, meta, dispatch, event);
      expect(mouseTool.onMouse).not.toHaveBeenCalled();
    });

    it('dispatches a MOUSE action', () => {
      mouseTool.onMouseMove(state, meta, dispatch, event);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: MOUSE,
        position: { x: 7, y: 1 },
      });
    });
  });

  describe('mouse release', () => {
    const position = {};
    const event = {};

    const state = {
      mouse: {
        startingPosition: position,
      },
    };

    it('does not call onMouseRelease if there was no onMousePress', () => {
      mouseTool.onMouseUp({ mouse: {} }, meta, dispatch);
      expect(mouseTool.onMouseRelease).not.toHaveBeenCalled();
    });

    it('calls onMouseRelease if there was a onMousePress', () => {
      mouseTool.onMouseUp(state, meta, dispatch, event);
      expect(mouseTool.onMouseRelease).toHaveBeenCalledWith(state.mouse, undefined, dispatch, event);
    });

    it('dispatches a MOUSE_RELEASE action', () => {
      mouseTool.onMouseUp(state, meta, dispatch);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: MOUSE_RELEASE,
      });
    });
  });
});
