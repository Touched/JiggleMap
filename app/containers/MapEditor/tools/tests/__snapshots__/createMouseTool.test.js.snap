// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMouseTool reducer sets the intial state 1`] = `
Object {
  "mouse": Object {
    "clientOffset": null,
    "currentBlock": 0,
    "currentPosition": null,
    "object": null,
    "startingPosition": null,
  },
  "tool": Object {
    "foo": "bar",
  },
}
`;

exports[`createMouseTool reducer sets the intial state 2`] = `
Object {
  "mouse": Object {
    "clientOffset": null,
    "currentBlock": 0,
    "currentPosition": null,
    "object": null,
    "startingPosition": null,
  },
  "tool": Object {},
}
`;
